telegram/_stars.py

```diff
@@ -0,0 +1,426 @@
+        type (:obj:`str`): The type of the transaction partner,
+            always :tg-const:`telegram.TransactionPartner.FRAGMENT`.
+        withdrawal_state (:obj:`telegram.RevenueWithdrawalState`): Optional. State of the
+            transaction if the transaction is outgoing.
+    """
+
+    __slots__ = ("withdrawal_state",)
+
+    def __init__(
+        self,
+        withdrawal_state: Optional["RevenueWithdrawalState"] = None,
+        *,
+        api_kwargs: Optional[JSONDict] = None,
+    ) -> None:
+        super().__init__(type=TransactionPartner.FRAGMENT, api_kwargs=api_kwargs)
+
+        with self._unfrozen():
+            self.withdrawal_state: Optional[RevenueWithdrawalState] = withdrawal_state
+
+    @classmethod
+    def de_json(
+        cls, data: Optional[JSONDict], bot: "Bot"
+    ) -> Optional["TransactionPartnerFragment"]:
+        data = cls._parse_data(data)
+
+        if not data:
+            return None
+
+        data["withdrawal_state"] = RevenueWithdrawalState.de_json(
+            data.get("withdrawal_state"), bot
+        )
+
+        return super().de_json(data=data, bot=bot)  # type: ignore[return-value]
+
+
+class TransactionPartnerUser(TransactionPartner):
+    """Describes a transaction with a user.
+
+    .. versionadded:: NEXT.VERSION
+
+    Args:
+        user (:class:`telegram.User`): Information about the user.
+
+    Attributes:
+        type (:obj:`str`): The type of the transaction partner,
+            always :tg-const:`telegram.TransactionPartner.USER`.
+        user (:class:`telegram.User`): Information about the user.
+    """
+
+    __slots__ = ("user",)
+
+    def __init__(self, user: "User", *, api_kwargs: Optional[JSONDict] = None) -> None:
+        super().__init__(type=TransactionPartner.USER, api_kwargs=api_kwargs)
+
+        with self._unfrozen():
+            self.user: User = user
+
+    @classmethod
+    def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["TransactionPartnerUser"]:
+        data = cls._parse_data(data)
+
+        if not data:
+            return None
+
+        data["user"] = User.de_json(data.get("user"), bot)
+
+        return super().de_json(data=data, bot=bot)  # type: ignore[return-value]
+
+
+class TransactionPartnerOther(TransactionPartner):
+    """Describes a transaction with an unknown partner.
+
+    .. versionadded:: NEXT.VERSION
+
+    Attributes:
+        type (:obj:`str`): The type of the transaction partner,
+            always :tg-const:`telegram.TransactionPartner.OTHER`.
+    """
+
+    __slots__ = ()
+
+    def __init__(self, *, api_kwargs: Optional[JSONDict] = None) -> None:
+        super().__init__(type=TransactionPartner.OTHER, api_kwargs=api_kwargs)
+
+
+class StarTransaction(TelegramObject):
+    """Describes a Telegram Star transaction.
+
+    Objects of this class are comparable in terms of equality. Two objects of this class are
+    considered equal, if their :attr:`id`, :attr:`amount`, and :attr:`date` are equal.
```

Comment:
looking at the attribute descriptions I would have selected id, source, receiver. My reasoning:

* id is unique up to describing if its a original transaction or a refund
* the latter is identified by source and receiver

Reply:
I just use required arguments by default for comparison, source and receiver are optional so do you still want that?

Reply:
using required arguments is the sane thing to do in most cases, true, but IMO more meaningful comparison overrules that. Plus, for mutually exclusive optionals, using all of them is a pattern that we already use e.g. in IKB. If my reasoning makes sense to you, then I'd like to go that way :)